src/parquet/column/column-reader-test.cc

```diff
@@ -44,195 +44,135 @@ namespace test {
 
 class TestPrimitiveReader : public ::testing::Test {
  public:
-  void SetUp() {}
-
-  void TearDown() {}
+  void make_pages() {
+    num_values_ = 0;
+    int levels_per_page = num_levels_ / num_pages_;
+    vector<int16_t> defs;
+    vector<int16_t> reps;
+    vector<int32_t> vals;
+    int values_per_page = 0;
+    for (int p = 0; p < num_pages_; p++) {
+      values_per_page = 0;
+      // Create definition levels
+      if (max_def_level_ > 0) {
+        defs.resize(levels_per_page);
+        random_levels(levels_per_page, 0, max_def_level_, defs.data());
+        for (int i = 0; i < levels_per_page; i++) {
+          if (defs[i] == max_def_level_) values_per_page++;
+        }
+        def_levels_.insert(def_levels_.end(), defs.begin(), defs.end());
+      } else {
+        values_per_page = levels_per_page;
+      }
+      // Create repitition levels
+      if (max_rep_level_ > 0) {
+        reps.resize(levels_per_page);
+        random_levels(levels_per_page, 0, max_rep_level_, reps.data());
+        rep_levels_.insert(rep_levels_.end(), reps.begin(), reps.end());
+      }
+      // Create values
+      vals.resize(values_per_page);
+      random_numbers(values_per_page, 0, vals.data());
+      values_.insert(values_.end(), vals.begin(), vals.end());
+      std::shared_ptr<DataPage> page = MakeDataPage<Type::INT32>(vals, defs,
+          max_def_level_, reps, max_rep_level_);
+      pages_.push_back(page);
+
+      num_values_ += values_per_page;
+    }
+  }
 
-  void InitReader(const ColumnDescriptor* descr) {
+  void init_reader() {
     pager_.reset(new test::MockPageReader(pages_));
-    reader_ = ColumnReader::Make(descr, std::move(pager_));
+    reader_ = ColumnReader::Make(descr_, std::move(pager_));
+  }
+
+  void check_results() {
+    vector<int32_t> vresult(num_values_, -1);
+    vector<int16_t> dresult(num_levels_, -1);
+    vector<int16_t> rresult(num_levels_, -1);
+    size_t values_read = 0;
+    size_t total_values_read = 0;
+    size_t batch_actual = 0;
+
+    Int32Reader* reader = static_cast<Int32Reader*>(reader_.get());
+    int32_t batch_size = 8;
+    size_t batch = 0;
+    // This will cover both the cases
+    // 1) batch_size < page_size (multiple ReadBatch from a single page)
+    // 2) batch_size > page_size (BatchRead from multiple pages)
+    do {
+      batch = reader->ReadBatch(batch_size, &dresult[0] + batch_actual,
+          &rresult[0] + batch_actual, &vresult[0] + total_values_read, &values_read);
```

Comment:
I think it's only an issue for compressed data pages (because of the `decompression_buffer_`). I think until we figure out a more flexible way to manage decompressed data lifetime we should return the batch when we reach the end of the data page, and document that users should keep calling `ReadBatch` until it returns 0